service/history/replicationTaskExecutor.go

```diff
@@ -0,0 +1,342 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination replicationTaskExecutor_mock.go -self_package github.com/uber/cadence/service/history
+
+package history
+
+import (
+	"context"
+
+	"github.com/uber/cadence/.gen/go/history"
+	r "github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/xdc"
+)
+
+type (
+	replicationTaskExecutor interface {
+		execute(sourceCluster string, replicationTask *r.ReplicationTask, forceApply bool) (int, error)
+	}
+
+	replicationTaskExecutorImpl struct {
+		currentCluster      string
+		domainCache         cache.DomainCache
+		nDCHistoryResender  xdc.NDCHistoryResender
+		historyRereplicator xdc.HistoryRereplicator
+		historyEngine       Engine
+
+		metricsClient metrics.Client
+		logger        log.Logger
+	}
+)
+
+func newReplicationTaskExecutor(
+	currentCluster string,
+	domainCache cache.DomainCache,
+	nDCHistoryResender xdc.NDCHistoryResender,
+	historyRereplicator xdc.HistoryRereplicator,
+	historyEngine Engine,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) replicationTaskExecutor {
+	return &replicationTaskExecutorImpl{
+		currentCluster:      currentCluster,
+		domainCache:         domainCache,
+		nDCHistoryResender:  nDCHistoryResender,
+		historyRereplicator: historyRereplicator,
+		historyEngine:       historyEngine,
+		metricsClient:       metricsClient,
+		logger:              logger,
+	}
+}
+
+func (e *replicationTaskExecutorImpl) execute(
+	sourceCluster string,
+	replicationTask *r.ReplicationTask,
+	forceApply bool,
+) (int, error) {
+
+	var err error
+	var scope int
+	switch replicationTask.GetTaskType() {
+	case r.ReplicationTaskTypeDomain:
+		// Domain replication task should be handled in worker (domainReplicationMessageProcessor)
+		panic("task type not supported")
```

Comment:
does panicking help ? I would suggest gracefully handling this by emitting log / metrics and skipping the task.

Reply:
Removed the case and should be handled by the default